base/compiler/tfuncs.jl

```diff
@@ -1699,6 +1699,72 @@ function _builtin_nothrow(@nospecialize(f), argtypes::Array{Any,1}, @nospecializ
     return false
 end
 
+# known to be always effect-free (in particular nothrow)
+const _PURE_BUILTINS = Any[tuple, svec, ===, typeof, nfields]
+
+# known to be effect-free if the are nothrow
+const _PURE_OR_ERROR_BUILTINS = [
+    fieldtype, apply_type, isa, UnionAll,
+    getfield, arrayref, const_arrayref, isdefined, Core.sizeof,
+    Core.kwfunc, Core.ifelse, Core._typevar, (<:),
```

Comment:
Looks like `arraysize` is missing here?

Reply:
This list is just moved. However, as defined `arraysize` doesn't go in this list, because the old notion of purity includes `:consistent`. In the new terminology, arraysize is `:effect_free`, but not `:consistent`.

Reply:
I didn't go though all the modifications. But I guess the following is not intended?
```julia
julia> Core.Compiler.builtin_effects(Base.arraysize, Any[Vector{Int},Int], Int)
(!c,!e,?n,+t)
```

Reply:
Yes, it should be `(!c, +e, ?n, +t)`. Actually looking at where uses of `_PURE_OR_ERROR_BUILTINS` remain, you are probably correct that we only use it in the `:effect_free` sense at this point, so we could probably rename it `_EFFECT_FREE_BUILTINS` and add `arraysize`.